boards: tcc: topst_vcp45: add Telechips' TOPST-VCP45 board and basic device driver

pooky1004 opened the pull request:
- Add vendor-prefix tcc for vendor TELECHIPS INC.
- Add support for TCC7045, SoC of TOPST VCP45.
- Add device-tree files
- Add support for TOPST VCP45 board
- Add GPIO driver for tccvcp 
- Add clock-control driver
- Add interrupt-controller driver
- Add timer driver
- Add serial driver
- Add console driver
- Add trigger_irq for topst_vcp45

nordicjm commented on drivers/interrupt_controller/Kconfig.tic:
```diff
@@ -0,0 +1,14 @@
+	default y
+	depends on DT_HAS_TCC_TIC_ENABLED
+	help
+		The Telechips Interrupt Controller provides hardware assistance for prioritizing
+		and aggregating the interrupt sources for ARM Cortex-R5 processor cores.
```
help text indent it 1x tab followed by 2x spaces

pooky1004 replied:
fix help text indentation in drivers/interrupt controller/Kconfig.tic file, replace 2x tabs with 1x tab followed by 2x spaces.

nordicjm commented on include/zephyr/drivers/clock_control/clock_control_tcc_ccu.h:
```diff
@@ -0,0 +1,53 @@
+
+void vcp_clock_init(void);
+
+long clock_set_pll_rate(signed long id, uint32_t rate);
```
@stephanosio do we allow long types instead of int?

pooky1004 replied:
change signed long and long to int32_t, unsigned long long to uint64_t, and boolean to bool.

nordicjm commented on include/zephyr/drivers/clock_control/clock_control_tcc_ccu.h:
```diff
@@ -0,0 +1,53 @@
+
+long clock_set_pll_rate(signed long id, uint32_t rate);
+
+uint32_t clock_get_pll_rate(signed long id);
```
None of these are documented, doesn't look to be using the zephyr driver class properly @nordic-krch is this acceptable?

nordic-krch replied:
`include/zephyr/drivers/clock_control` contains vendor specific extensions to clock related features so that is ok but definitely it is missing doxygen documentation.

nordicjm commented on include/zephyr/drivers/clock_control/clock_control_tcc_ccu.h:
```diff
@@ -0,0 +1,53 @@
+#ifndef CLOCK_CONTROL_TCC_CCU_HEADER
+#define CLOCK_CONTROL_TCC_CCU_HEADER
+
+/*
+ * DEFINITIONS
+ *
+ */
```
```suggestion
```

pooky1004 replied:
I don't know the difference between the two. There appears to be no difference when I copy and paste the part you suggested.

nordicjm commented on include/zephyr/drivers/clock_control/clock_control_tcc_ccu.h:
```diff
@@ -0,0 +1,53 @@
+
+#define CKC_MHZ(x) (uint32_t)((x) * 1000000)
+
+/*
+ * FUNCTION PROTOTYPES
+ *
+ */
```
```suggestion
```
thee headers can go

nordicjm commented on include/zephyr/drivers/gpio/gpio_tccvcp.h:
```diff
@@ -0,0 +1,203 @@
+ *
+ * Notes
+ */
+int32_t vcp_gpio_config(uint32_t uiPort, uint32_t uiConfig);
```
snake case not camel case

nordicjm commented on boards/tcc/topst_vcp45/Kconfig.defconfig:
```diff
@@ -0,0 +1,88 @@
+config ARM_CUSTOM_INTERRUPT_CONTROLLER
+	default y
+
+config THREAD_MONITOR
+	default y
+
+config THREAD_STACK_INFO
+	default y
+
+config INIT_STACKS
+	default y
+
+config TRACING
+	default y
+
+config TRACING_OBJECT_TRACKING
+	default y
+
+config TRACING_SYSCALL
+	default y
+
+config TRACING_THREAD
+	default y
+
+config TRACING_WORK
+	default y
+
+config TRACING_ISR
+	default y
```
so why do any of these need to be enabled? Include for all the options below too

nordicjm commented on boards/tcc/topst_vcp45/board.cmake:
```diff
@@ -0,0 +1,3 @@
+# Copyright 2024 Hounjoung Rim <hounjoung@tsnlab.com>
+#
+# SPDX-License-Identifier: Apache-2.0
```
empty file can go

nordicjm commented on boards/tcc/topst_vcp45/doc/index.rst:
```diff
@@ -0,0 +1,506 @@
+   :board: topst_vcp45
+   :goals: topst_vcp45
+
+Change all 'printf' in the samples/hello_world/src/main.c file to 'printk'.
```
why?

nordicjm commented on boards/tcc/topst_vcp45/doc/index.rst:
```diff
@@ -0,0 +1,506 @@
+
+Change all 'printf' in the samples/hello_world/src/main.c file to 'printk'.
+
+west build --build-dir topst_vcp45 -b topst_vcp45 samples/hello_world
```
this needs to use the proper zephyr code directive tag

nordicjm replied:
not addressed

nordicjm replied:
not addressed

nordicjm commented on boards/tcc/topst_vcp45/doc/topst_vcp45.png:
convert to webp then put through https://tinypng.com/

nordicjm commented on boards/tcc/topst_vcp45/topst_vcp45_defconfig:
```diff
@@ -0,0 +1,31 @@
+CONFIG_UART_CONSOLE_MCUMGR=n
+CONFIG_CBPRINTF_LIBC_SUBSTS=y
+
+CONFIG_SYS_CLOCK_TICKS_PER_SEC=1000
```
this should be done in soc

nordicjm commented on boards/tcc/topst_vcp45/topst_vcp45_defconfig:
```diff
@@ -0,0 +1,31 @@
+
+CONFIG_SYSCON=y
+
+CONFIG_TIMEOUT_64BIT=y
```
soc?

pooky1004 replied:
move some configs from boards/tcc/topst_vcp45/topst_vcp45_defconfig to SOC config

nordicjm commented on boards/tcc/topst_vcp45/topst_vcp45_defconfig:
```diff
@@ -0,0 +1,31 @@
+# Copyright 2024 Hounjoung Rim <hounjoung@tsnlab.com>
+
+CONFIG_XIP=n
+CONFIG_SRAM_VECTOR_TABLE=y
```
soc?

pooky1004 replied:
move some configs from boards/tcc/topst_vcp45/topst_vcp45_defconfig to SOC config

nordicjm commented on boards/tcc/topst_vcp45/topst_vcp45_defconfig:
```diff
@@ -0,0 +1,31 @@
+CONFIG_XIP=n
+CONFIG_SRAM_VECTOR_TABLE=y
+
+CONFIG_MAX_THREAD_BYTES=3
+CONFIG_THREAD_STACK_INFO=y
+
```
```suggestion
```

pooky1004 replied:
change the value of CONFIG_MAX_THREAD_BYTES defined in the boards/tcc/topst_vcp45/topst_vcp45_defconfig file from 3 to 13.

nordicjm commented on soc/tcc/tcc7045/linker.ld:
```diff
@@ -0,0 +1,385 @@
+/*
+ * Copyright 2024 Hounjoung Rim <hounjoung@tsnlab.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
```
why is this needed and why can't one of the cortex_* ones in tree be used?

pooky1004 replied:
Additional settings are needed for REMAP, so I use our own linker.ld file.

nordicjm replied:
@tejlmand

tejlmand replied:
Sometimes SoC's needs there own linker.ld file for various reason, and the maintainer or contributor of this should be ready to update when there are changes.

Though if the sole reason in this case is for a `REMAP` memory section, then I would like to know why this memory region is not defined in devicetree and thereby become available through the use of `LINKER_DT_REGIONS()`, which also exists in the cortex default linker templates.

Based on this observation, then I also doubt this SoC works with the CMake linker generator.
A basic test of this would simply be to do a build using `-DCONFIG_CMAKE_LINKER_GENERATOR=y`.
(Haven't tested this PR yet)

nordicjm commented on soc/tcc/tcc7045/soc.c:
```diff
@@ -0,0 +1,154 @@
+int32_t soc_div64_to_32(unsigned long long *dividend_ptr, uint32_t divisor, uint32_t *rem_ptr)
+{
+	int32_t ret_val = 0;
+	unsigned long long rem = 0;
+	unsigned long long b = divisor;
+	unsigned long long d = 1;
+	unsigned long long res = 0;
```
why can't uint64 be used?

pooky1004 replied:
change signed long and long to int32_t, unsigned long long to uint64_t, and boolean to bool.

nordicjm commented on soc/tcc/tcc7045/soc.h:
```diff
@@ -0,0 +1,46 @@
+
+int32_t soc_div64_to_32(unsigned long long *pullDividend, uint32_t uiDivisor, uint32_t *puiRem);
+
+typedef unsigned char boolean; /* for use with TRUE/FALSE        */
```
no

pooky1004 replied:
change signed long and long to int32_t, unsigned long long to uint64_t, and boolean to bool.

nordicjm commented on soc/tcc/tcc7045/soc.h:
```diff
@@ -0,0 +1,46 @@
+#ifndef _BOARD_SOC__H_
+#define _BOARD_SOC__H_
+
+#ifndef FALSE
+#define FALSE (0U)
+#endif
+
+#ifndef TRUE
+#define TRUE (1U)
+#endif
+
+#ifndef NULL_PTR
+#define NULL_PTR ((void *)0)
+#endif
+
+#ifndef NULL
+#define NULL (0)
+#endif
+
+#ifndef ON
+#define ON  (TRUE)
+#define OFF (FALSE)
+#endif
```
no

rruuaanng commented on include/zephyr/dt-bindings/interrupt-controller/tcc-tic.h:
```diff
@@ -0,0 +1,158 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef TCC_TIC_ENUM_HEADER
+#define TCC_TIC_ENUM_HEADER
```
```suggestion
#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_INTERRUPT_CONTROLLER_TCC_TIC_H_
#define ZEPHYR_INCLUDE_DT_BINDINGS_INTERRUPT_CONTROLLER_TCC_TIC_H_
```

rruuaanng commented on include/zephyr/dt-bindings/interrupt-controller/tcc-tic.h:
```diff
@@ -0,0 +1,158 @@
+
+#define TIC_INT_SRC_CNT (TIC_SPU + 1UL)
+
+#endif /* TCC_TIC_ENUM_HEADER */
```
```suggestion
#endif /* ZEPHYR_INCLUDE_DT_BINDINGS_INTERRUPT_CONTROLLER_TCC_TIC_H_ */
```

rruuaanng reviewed (changes requested):
I'm glad I didn't arrive too late. As for the preprocessing macros, we should align with Zephyr. Regarding the other header files, I think they also need to be updated, but I won't point them out here.

rruuaanng commented on include/zephyr/dt-bindings/interrupt-controller/tcc-tic.h:
```diff
@@ -0,0 +1,158 @@
+ * DEFINITIONS
+ *
+ */
+
```
We should remove these brackets, as they are redundant (to see other related files.), like this
```
#define TIC_INT_TYPE_LEVEL_HIGH 0x1U
```
not
```
#define TIC_INT_TYPE_LEVEL_HIGH (0x1U)
```

e.g.
```
from clock

#define STM32_CLOCK_BUS_IOP     0x034
#define STM32_CLOCK_BUS_AHB1    0x038

from adc

#define DT_ADC_GPIO_PB2      0x03
#define DT_ADC_GPIO_PB3      0x04
#define DT_ADC_GPIO_PB4      0x05
#define DT_ADC_GPIO_PB5      0x06
```

rruuaanng commented on include/zephyr/dt-bindings/interrupt-controller/tcc-tic.h:
```diff
@@ -0,0 +1,158 @@
+#define TIC_EXT7                  (TIC_SPI_START + 78UL)
+#define TIC_EXT8                  (TIC_SPI_START + 79UL)
+#define TIC_EXT9                  (TIC_SPI_START + 80UL)
+#define TIC_EXTn0                 (TIC_SPI_START + 81UL)
+#define TIC_EXTn1                 (TIC_SPI_START + 82UL)
+#define TIC_EXTn2                 (TIC_SPI_START + 83UL)
+#define TIC_EXTn3                 (TIC_SPI_START + 84UL)
+#define TIC_EXTn4                 (TIC_SPI_START + 85UL)
+#define TIC_EXTn5                 (TIC_SPI_START + 86UL)
+#define TIC_EXTn6                 (TIC_SPI_START + 87UL)
+#define TIC_EXTn7                 (TIC_SPI_START + 88UL)
+#define TIC_EXTn8                 (TIC_SPI_START + 89UL)
+#define TIC_EXTn9                 (TIC_SPI_START + 90UL)
```
```suggestion
#define TIC_EXT_N0                 (TIC_SPI_START + 81UL)
#define TIC_EXT_N1                 (TIC_SPI_START + 82UL)
#define TIC_EXT_N2                 (TIC_SPI_START + 83UL)
#define TIC_EXT_N3                 (TIC_SPI_START + 84UL)
#define TIC_EXT_N4                 (TIC_SPI_START + 85UL)
#define TIC_EXT_N5                 (TIC_SPI_START + 86UL)
#define TIC_EXT_N6                 (TIC_SPI_START + 87UL)
#define TIC_EXT_N7                 (TIC_SPI_START + 88UL)
#define TIC_EXT_N8                 (TIC_SPI_START + 89UL)
#define TIC_EXT_N9                 (TIC_SPI_START + 90UL)
```

rruuaanng commented on include/zephyr/dt-bindings/interrupt-controller/tcc-tic.h:
```diff
@@ -0,0 +1,158 @@
+
+#define TIC_EINT_START_INT (TIC_EXT0)
+#define TIC_EINT_END_INT   (TIC_EXT9)
+#define TIC_EINT_NUM       ((unsigned long)TIC_EXTn0 - (unsigned long)TIC_EXT0)
```
```suggestion
#define TIC_EINT_NUM       ((unsigned long)TIC_EXT_N0 - (unsigned long)TIC_EXT0)
```
(Please forgive me, but I don't understand why the conversion is necessary. It seems redundant.)

rruuaanng commented on include/zephyr/drivers/clock_control/clock_control_tcc_ccu.h:
```diff
@@ -0,0 +1,51 @@
+ * DEFINITIONS
+ */
+
+#define CKC_MHZ(x) (uint32_t)((x) * 1000000)
```
Please remove it, we have `DT_FREQ_M`.
(or you can define an alias for it. It can be called `CKC_MHZ`.)

pooky1004 replied:
delete the unused CKC_MHZ(x) macro defined in the include/zephyr/drivers/clock_control/clock_control_tcc_ccu.h file.

rruuaanng reviewed (dismissed):
We should now address the other maintainers' comments and get the CI back to normal.
(Maybe I missed some parts, but I need to leave it to others.)

nordicjm commented on soc/tcc/tcc7045/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+
+zephyr_include_directories(.)
+
+zephyr_ld_options(--cref -Bstatic -nostdlib -nostartfiles -p -EL)
```
???

pooky1004 replied:
I already answered, Additional settings are needed for REMAP, so I use our own linker.ld file.

nordicjm commented on dts/arm/tcc/tcc70xx.dtsi:
```diff
@@ -0,0 +1,240 @@
+
+/ {
+	soc {
+
```
```suggestion
```

nordicjm commented on boards/tcc/topst_vcp45/Kconfig.defconfig:
```diff
@@ -0,0 +1,19 @@
+
+if BOARD_TOPST_VCP45
+
+config ARM_CUSTOM_INTERRUPT_CONTROLLER
+	default y
```
how is this board related?

pooky1004 replied:
I haven't found any code in zephyr for the interrupt controller we use on our board. So I wrote my own interrupt controller.

nordicjm commented on boards/tcc/topst_vcp45/Kconfig.defconfig:
```diff
@@ -0,0 +1,19 @@
+
+if USERSPACE
+
+config COMPILER_ISA_THUMB2
+	default n
```
and this?

nordicjm reviewed (commented):
why was my review re-requested when you haven't fixed or replied to the comments I originally left?

pooky1004 commented:
Sorry for the late reply.
Currently, comments cannot be made on the items below.

Comment on the items below in an email reply.

comment:
I was going to use printk as an example at the device driver level, but I
deleted that phrase because it didn't seem to matter.

In boards/tcc/topst_vcp45/doc/index.rst
<https://github.com/zephyrproject-rtos/zephyr/pull/83136#discussion_r1959268573>
:

> +hello_world
+===========
+
+zephyr.bin
+----------
+
+Build an app, for example :zephyr:code-sample:`hello_world`
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/hello_world
+   :board: topst_vcp45
+   :goals: topst_vcp45
+
+Change all 'printf' in the samples/hello_world/src/main.c file to 'printk'.
+
+west build --build-dir topst_vcp45 -b topst_vcp45 samples/hello_world

not addressed

sincerely


HounJoung Rim



2025년 2월 18일 (화) 오후 5:24, Jamie ***@***.***>님이 작성:

> ***@***.**** commented on this pull request.
>
> why was my review re-requested when you haven't fixed or replied to the
> comments I originally left?
> ------------------------------
>
> In soc/tcc/tcc7045/CMakeLists.txt
> <https://github.com/zephyrproject-rtos/zephyr/pull/83136#discussion_r1959262602>
> :
>
> > +# Copyright 2024 Hounjoung Rim ***@***.***>
> +#
> +# SPDX-License-Identifier: Apache-2.0
> +
> +zephyr_sources(
> +	soc.c
> +)
> +
> +zephyr_sources_ifdef(
> +  CONFIG_ARM_MPU
> +  arm_mpu_regions.c
> +)
> +
> +zephyr_include_directories(.)
> +
> +zephyr_ld_options(--cref -Bstatic -nostdlib -nostartfiles -p -EL)
>
> ???
> ------------------------------
>
> In dts/arm/tcc/tcc70xx.dtsi
> <https://github.com/zephyrproject-rtos/zephyr/pull/83136#discussion_r1959266955>
> :
>
> > @@ -0,0 +1,240 @@
> +/*
> + * Copyright 2024 Hounjoung Rim ***@***.***>
> + *
> + * SPDX-License-Identifier: Apache-2.0
> + */
> +
> +#include <mem.h>
> +#include <freq.h>
> +#include <arm/armv7-r.dtsi>
> +#include <zephyr/dt-bindings/interrupt-controller/tcc-tic.h>
> +
> +/ {
> +	soc {
> +
>
> ⬇️ Suggested change
>
> -
>
> ------------------------------
>
> In boards/tcc/topst_vcp45/Kconfig.defconfig
> <https://github.com/zephyrproject-rtos/zephyr/pull/83136#discussion_r1959267976>
> :
>
> > +config ARM_CUSTOM_INTERRUPT_CONTROLLER
> +	default y
>
> how is this board related?
> ------------------------------
>
> In boards/tcc/topst_vcp45/Kconfig.defconfig
> <https://github.com/zephyrproject-rtos/zephyr/pull/83136#discussion_r1959268099>
> :
>
> > +config COMPILER_ISA_THUMB2
> +	default n
>
> and this?
> ------------------------------
>
> In boards/tcc/topst_vcp45/doc/index.rst
> <https://github.com/zephyrproject-rtos/zephyr/pull/83136#discussion_r1959268573>
> :
>
> > +hello_world
> +===========
> +
> +zephyr.bin
> +----------
> +
> +Build an app, for example :zephyr:code-sample:`hello_world`
> +
> +.. zephyr-app-commands::
> +   :zephyr-app: samples/hello_world
> +   :board: topst_vcp45
> +   :goals: topst_vcp45
> +
> +Change all 'printf' in the samples/hello_world/src/main.c file to 'printk'.
> +
> +west build --build-dir topst_vcp45 -b topst_vcp45 samples/hello_world
>
> not addressed
> ------------------------------
>
> In soc/tcc/tcc7045/linker.ld
> <https://github.com/zephyrproject-rtos/zephyr/pull/83136#discussion_r1959265681>
> :
>
> > @@ -0,0 +1,385 @@
> +/*
> + * Copyright 2024 Hounjoung Rim ***@***.***>
> + *
> + * SPDX-License-Identifier: Apache-2.0
>
> @tejlmand <https://github.com/tejlmand>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/zephyrproject-rtos/zephyr/pull/83136#pullrequestreview-2622757655>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APFAC7QXHZBDPQOZFWJPFQ32QLU5BAVCNFSM6AAAAABTZTCKACVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMMRSG42TONRVGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

임헌정 / HounJoung Rim

티에스엔랩(주) / TSN Lab, Inc.

Address: 경기도 용인시 기흥구 기흥로 60-1, 1002호

Mail: ***@***.*** ***@***.***>

Tel: +31-282-8290

Mobile: +82-10-9530-2342

nordicjm commented on drivers/clock_control/clock_control_tccvcp.c:
```diff
@@ -0,0 +1,1127 @@
+
+#include <soc_reg_phys.h>
+
+/*
+ * DEFINITIONS
+ *
+ */
+
+struct clock_tcc_config {
+	const struct device *syscon;
+};
+
+#define DEV_CFG(dev) ((const struct clock_tcc_config *const)(dev)->config)
+
+/*
+ * LOCAL VARIABLES
+ *
+ */
+
+static uint32_t micom_clock_source[CLOCK_SRC_MAX_NUM];
+
+/*
+ * FUNCTION PROTOTYPES
+ *
+ */
```
do we really need these headers?

nordicjm commented on dts/arm/tcc/tcc7045.dtsi:
```diff
@@ -0,0 +1,33 @@
+	};
+
+	soc {
+
```
```suggestion
```

nordicjm commented on dts/arm/tcc/tcc70xx.dtsi:
```diff
@@ -0,0 +1,239 @@
+		syscon: syscon@a0f24000 {
+			compatible = "syscon";
+			reg = <0xa0f24000 0x104>;
+			sysclk: sysclk {
```
newline

nordicjm replied:
not resolved

nordicjm commented on soc/tcc/tcc7045/Kconfig.defconfig:
```diff
@@ -0,0 +1,32 @@
+	def_bool y
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 12000000
```
this should come from dts

nordicjm commented on dts/arm/tcc/tcc7045.dtsi:
```diff
@@ -0,0 +1,30 @@
+			compatible = "arm,cortex-r5f";
+			reg = <0>;
+		};
+		tic: interrupt-controller@a0f10000 {
```
newline

nordicjm commented on soc/tcc/tcc7045/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+# Copyright 2024 Hounjoung Rim <hounjoung@tsnlab.com>
+#
+# SPDX-License-Identifier: Apache-2.0
```
soc commit must come before board commit

nordicjm commented on boards/tcc/topst_vcp45/doc/index.rst:
```diff
@@ -0,0 +1,504 @@
+Supported Features
+==================
+
+The following features are supported:
```
@kartben

nordicjm commented on soc/tcc/tcc7045/Kconfig.defconfig:
```diff
@@ -0,0 +1,44 @@
+config FLASH_BASE_ADDRESS
+	default $(dt_chosen_reg_addr_hex,$(DT_CHOSEN_Z_FLASH))
+
+config INIT_STACKS
+	default y
+
```
why is this needed?

nordicjm commented on soc/tcc/tcc7045/Kconfig.defconfig:
```diff
@@ -0,0 +1,44 @@
+config ARM_MPU
+	def_bool y
+
+config REQUIRES_FULL_LIBC
+	def_bool y
```
why is this needed?

nordicjm commented on soc/tcc/tcc7045/Kconfig.defconfig:
```diff
@@ -0,0 +1,44 @@
+config TICKLESS_KERNEL
+	def_bool n
+
+config TCCVCP_CONSOLE
+	def_bool y
```
this isn't even used, remove

nordicjm reviewed (changes requested):
You need to fix the issues that are listed, stop re-requesting my review again until they are **all** addressed

Commit:
dts: bindings: add vendor-prefix tcc for vendor TELECHIPS INC.

Add vendor-prefix tcc for vendor TELECHIPS INC., required by TOPST VCP45.

Signed-off-by: Hounjoung Rim <hounjoung@tsnlab.com>

Commit:
soc: tcc: add support for TCC7045

Add support for TCC7045, SoC of TOPST VCP45.

Signed-off-by: Hounjoung Rim <hounjoung@tsnlab.com>

Commit:
dts: arm: tcc: add device-tree for TOPST VCP45

Add device-tree files, required by TOPST VCP45.
The device tree defines hardware details such as the CPU, timers,
interrupt controller, flash memory, RAM, clock controller, UART,
and GPIOs, providing a clear map of the system's architecture.

Signed-off-by: Hounjoung Rim <hounjoung@tsnlab.com>

Commit:
board: tcc: add support for TOPST VCP45

Add support for TOPST VCP45.

Signed-off-by: Hounjoung Rim <hounjoung@tsnlab.com>

Commit:
drivers: gpio: add tccvcp gpio driver

Add GPIO driver for tccvcp, required by TOPST VCP45.

Signed-off-by: Hounjoung Rim <hounjoung@tsnlab.com>

Commit:
drivers: interrupt_controller: add tic interrupt-controller driver

Add interrupt-controller driver for tic, required by TOPST VCP45.

Signed-off-by: Hounjoung Rim <hounjoung@tsnlab.com>

Commit:
drivers: timer: add tcc_vcpttc timer driver

Add timer driver for tcc_vcpttc, required by TOPST VCP45.

Signed-off-by: Hounjoung Rim <hounjoung@tsnlab.com>

Commit:
drivers: clock_control: add tccvcp clock-control driver

Add clock-control driver for tccvcp, required by TOPST VCP45.

Signed-off-by: Hounjoung Rim <hounjoung@tsnlab.com>

Commit:
drivers: serial: add tccvcp serial driver

Add serial driver for tccvcp, required by TOPST VCP45.

Signed-off-by: Hounjoung Rim <hounjoung@tsnlab.com>

Commit:
drivers: console: add tccvcp console driver

Add console driver for tccvcp, required by TOPST VCP45.

Signed-off-by: Hounjoung Rim <hounjoung@tsnlab.com>

Commit:
drivers: interrupt_controller: subsys: add trigger_irq for topst_vcp45

add trigger_irq for topst vcp45 to
subsys/test/suite/include/zephyr/interrupt util.h file.

Signed-off-by: Hounjoung Rim <hounjoung@tsnlab.com>

stephanosio commented on include/zephyr/drivers/interrupt_controller/intc_tic.h:
```diff
@@ -0,0 +1,205 @@
+#include <zephyr/dt-bindings/interrupt-controller/tcc-tic.h>
+#include <zephyr/sys/util_macro.h>
+
+typedef struct tic_distributor {
```
This looks awful lot like ARM GICv1, if not identical ... Any reason this cannot use the existing GIC driver?
